sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java

```diff
@@ -0,0 +1,216 @@
+    service.load();
+
+    PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+    CassandraIO.Read spec = CassandraIO.read().withCassandraService(service);
+    CassandraIO.CassandraSource source = new CassandraIO.CassandraSource(
+        spec,
+        null);
+    long estimatedSizeBytes = source.getEstimatedSizeBytes(pipelineOptions);
+    assertEquals(113890, estimatedSizeBytes);
+  }
+
+  @Test
+  public void testRead() throws Exception {
+    FakeCassandraService service = new FakeCassandraService();
+    service.load();
+
+    PCollection<Scientist> output = (PCollection<Scientist>) pipeline.apply(CassandraIO
+        .<Scientist>read()
+        .withCassandraService(service)
+        .withKeyspace("beam")
+        .withTable("scientist")
+        .withCoder(SerializableCoder.of(Scientist.class))
+        .withEntity(Scientist.class)
+    );
+
+    PAssert.thatSingleton(output.apply("Count", Count.<Scientist>globally()))
+        .isEqualTo(10000L);
+
+    PCollection<KV<String, Integer>> mapped =
+        output.apply(
+            MapElements.via(
+                new SimpleFunction<Scientist, KV<String, Integer>>() {
+                  public KV<String, Integer> apply(Scientist scientist) {
+                    return KV.of(scientist.name, scientist.id);
+                  }
+                }));
+    PAssert.that(mapped.apply("Count occurrences per scientist", Count.<String, Integer>perKey()))
+        .satisfies(
+            new SerializableFunction<Iterable<KV<String, Long>>, Void>() {
+              @Override
+              public Void apply(Iterable<KV<String, Long>> input) {
+                for (KV<String, Long> element : input) {
+                  assertEquals(element.getKey(), 1000, element.getValue().longValue());
+                }
+                return null;
+              }
+            });
+
+    pipeline.run();
+  }
+
+  /**
+   * A {@link CassandraService} implementation that stores the entity in memory.
+   */
+  private static class FakeCassandraService implements CassandraService {
+
+    private static final Map<Integer, Scientist> table = new HashMap<>();
+
+    public void load() {
+      table.clear();
+      String[] scientists = {
+          "Lovelace",
+          "Franklin",
+          "Meitner",
+          "Hopper",
+          "Curie",
+          "Faraday",
+          "Newton",
+          "Bohr",
+          "Galilei",
+          "Maxwell"
+      };
+      for (int i = 0; i < 10000; i++) {
+        int index = i % scientists.length;
+        Scientist scientist = new Scientist();
+        scientist.id = i;
+        scientist.name = scientists[index];
+        table.put(scientist.id, scientist);
+      }
+    }
+
+    @Override
+    public Reader createReader(String query, Class entity) throws IOException {
+      return new FakeCassandraReader();
+    }
+
+    static class FakeCassandraReader implements CassandraService.Reader {
```

Comment:
I think you said you created the interface because it was easier to mock in tests. But this implementation looks pretty much exactly like an equivalent BoundedReader implementation would have looked like - you're overriding exactly the same methods as you'd be required to override there - so I still think that it's better to get rid of the interface and save on the amount of code.

Also, we should I guess have a utility implementation of a BoundedReader that reads a given fixed set of values - then you could reuse it here directly and further reduce the code. Actually, a PR to introduce such a utility would be very welcome :)